source/framework/vx_image.c

```diff
@@ -376,6 +381,310 @@ static vx_status ownAllocImageBuffer(vx_reference ref)
+    vx_rectangle_t rect = {.start_x = 0, .start_y = 0, .end_x = ip_objd->width, .end_y = ip_objd->height};
+    vx_imagepatch_addressing_t addr;
+    vx_map_id map_id;
+    void *ptr;
+    for (i = 0; i < ip_objd->planes; ++i)
+    {
+        if (ip_objd->mem_size[i] == op_objd->mem_size[i])
+        {
+            tivxCheckStatus(&status, tivxMemBufferMap((void *)(uintptr_t)ip_objd->mem_ptr[i].host_ptr, ip_objd->mem_size[i], 
+                                                      (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_READ_ONLY));
+            if ((vx_status)VX_SUCCESS == status)
+            {
+                tivxCheckStatus(&status, tivxMemBufferMap((void *)(uintptr_t)op_objd->mem_ptr[i].host_ptr, ip_objd->mem_size[i], 
+                                                          (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_WRITE_ONLY));
+                if ((vx_status)VX_SUCCESS == status)
+                {
+                    (void)memcpy((void *)(uintptr_t)op_objd->mem_ptr[i].host_ptr, (void *)(uintptr_t)ip_objd->mem_ptr[i].host_ptr, ip_objd->mem_size[i]);
+                }
+                tivxCheckStatus(&status, tivxMemBufferUnmap((void *)(uintptr_t)op_objd->mem_ptr[i].host_ptr, op_objd->mem_size[i],
+                                                            (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_WRITE_ONLY));
+            }
+            tivxCheckStatus(&status, tivxMemBufferMap((void *)(uintptr_t)ip_objd->mem_ptr[i].host_ptr, ip_objd->mem_size[i],
+                                                      (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_READ_ONLY));
+            if ((vx_status)VX_SUCCESS != status)
+            {
+                break;
+            }
+        }
+        else
+        {
+            tivxCheckStatus(&status, vxMapImagePatch(input, &rect, i, &map_id, &addr, &ptr, (vx_enum)VX_READ_ONLY, (vx_enum)VX_MEMORY_TYPE_HOST, 0));
+            if ((vx_status)VX_SUCCESS == status)
+            {
+                status = vxCopyImagePatch(output, &rect, i, &addr, ptr, (vx_enum)VX_WRITE_ONLY, (vx_enum)VX_MEMORY_TYPE_HOST);
+            }
+            tivxCheckStatus(&status, vxUnmapImagePatch(input, map_id));
+        }
+    }
+    return status;
+}
+
+/*! \brief Adjust the memory pointer of the tensor or image by the given offset
+ * This involves non-recursively visiting every sub-image and sub-tensor
+ * and adjusting the memory pointer for that, as well.
+ */
+static vx_status adjustMemoryPointer(vx_image ref, uint64_t offset[TIVX_IMAGE_MAX_PLANES])
+{
+    vx_status status = (vx_status)VX_SUCCESS;
+    /* the stack can hold the parent reference + the max number of subimages*/
+    vx_image stack[TIVX_IMAGE_MAX_SUBIMAGES+1];
+    vx_image local_img;
+    vx_image *subimages = NULL;
+    vx_int32 stack_pointer = 0;
+    vx_uint32 i;
+
+    for (i = 0; i < TIVX_IMAGE_MAX_SUBIMAGES; ++i)
+    {
+        stack[i] = NULL;
+    }
+    stack[stack_pointer++] = ref;
+    while (0U < stack_pointer)
```

Comment:
Let's discuss this function over a call.  I am not sure I am following it right.
If there are more than one level of sub-images, and the first level has maxed out the TIVX_IMAGE_MAX_SUBIMAGES value, then the lower levels will not have enough stack to iterate.  Should we just use TIVX_IMAGE_MAX_OBJECTS?  However, that number is big and too much for the local stack.  Maybe I am missing something, but we can talk when Lucas gets back.

Reply:
I agree with you, but in fact in the first version we had, we were iterating over TIVX_SUBOBJECT_STACK_SIZE which was 32 big. It got lost during the reworking with of the swap copy move feature because we had to remove the tensor from image stuff. TIVX_IMAGE_MAX_OBJECTS is too big, we only need 16 + 16 (2x the TIVX_IMAGE_MAX_SUBIMAGES)
can we create a new define ?

Reply:
@jessegv  and @lcweaver 
looking at the OVX specification, the number of level for the subimages, you have to support is not specified. So it is up to the implementation. Right now, I changed the stack size so that we can support 2 levels of subimages : 
image -> subimages -> subsubimages each with a size of max of subimages (16).
I also added a test which is testing this limit.
What do you think?

Reply:
@jessegv and @lcweaver  I had a short talk with Steve and we came to the conclusion that we have to restrict the depth of the subimage you can create. By doing that, you ensure to have a fixed stack size.
To be able to check the subimage depth at subimage creation (vxCreateImageFromROI and vxCreateImageFromChannel), we just have to check the parent hierarchy and refuse it if the depth is bigger than 2 for example (I think subimage from subimage should be sufficient).

Reply:
@lcweaver and I discussed this this afternoon.  I suggest to use a new #define that is defined only in this file at the top (not in the config files since it only affects local stack size and not object descriptor or context size): 

/* Max bound of iterating through sub-image for swapping; the size is directly proportional to the task stack usage */
#define TIVX_SUBIMAGE_STACK_SIZE (TIVX_IMAGE_MAX_SUBIMAGES*2U)

Then you can use it in this function and variable array size.

Reply:
done